core/domain/stats_services.py

```diff
@@ -352,76 +290,68 @@ def update_exp_issues_for_new_exp_version(
-
-        playthroughs = get_playthroughs_multi(playthrough_ids)
-        for p_idx, playthrough in enumerate(playthroughs):
-            if stats_models.CUSTOMIZATION_ARG_WHICH_IDENTIFIES_ISSUE[
-                    playthrough.issue_type] == 'state_names':
-                state_names = playthrough.issue_customization_args[
-                    'state_names']['value']
-                playthrough.issue_customization_args['state_names']['value'] = [
-                    new_state_name
-                    if state_name == old_state_name else state_name
-                    for state_name in state_names]
-            else:
-                playthrough.issue_customization_args['state_name']['value'] = (
-                    new_state_name)
-            for a_idx, action in enumerate(playthrough.actions):
-                if action.action_customization_args['state_name']['value'] == (
-                        old_state_name):
-                    playthroughs[p_idx].actions[
-                        a_idx].action_customization_args['state_name'][
-                            'value'] = new_state_name
-
-        all_playthrough_ids.extend(playthrough_ids)
-        all_playthroughs.extend(playthroughs)
-
-    update_playthroughs_multi(all_playthrough_ids, all_playthroughs)
+    deleted_state_names = exp_versions_diff.deleted_state_names
+    old_to_new_state_names = exp_versions_diff.old_to_new_state_names
+
+    playthrough_ids = list(itertools.chain.from_iterable(
+        i.playthrough_ids for i in exp_issues.unresolved_issues))
+    playthrough_models = (
+        stats_models.PlaythroughModel.get_multi(playthrough_ids))
+
+    for playthrough_model in playthrough_models:
+        playthrough = get_playthrough_from_model(playthrough_model)
+
+        if 'state_names' in playthrough.issue_customization_args:
+            state_names = (
+                playthrough.issue_customization_args['state_names']['value'])
+            playthrough.issue_customization_args['state_names']['value'] = [
+                state_name if state_name not in old_to_new_state_names else
+                old_to_new_state_names[state_name] for state_name in state_names
+            ]
+
+        if 'state_name' in playthrough.issue_customization_args:
+            state_name = (
+                playthrough.issue_customization_args['state_name']['value'])
+            playthrough.issue_customization_args['state_name']['value'] = (
+                state_name if state_name not in old_to_new_state_names else
+                old_to_new_state_names[state_name])
+
+        for action in playthrough.actions:
+            state_name = (
+                action.action_customization_args['state_name']['value'])
+            action.action_customization_args['state_name']['value'] = (
+                state_name if state_name not in old_to_new_state_names else
+                old_to_new_state_names[state_name])
+
+        playthrough_model.populate(**playthrough.to_dict())
```

Comment:
Is it always guaranteed that the playthrough structure will match the properties in the model? I feel like this is only true for some models in the codebase, and as the playthrough object changes this may have unintended side effects.

Reply:
Hmm it's guaranteed in the context of this PR, but it is indeed vulnerable to becoming out of sync. I could add a unit test case for protection, WDYT?

Reply:
I've decided to just document this expectation in the `to_dict` method, since the elegance is pretty nice.

Reply:
Obsolete-- I've made explicit assignments to each property instead.

Reply:
It seems like you're still using populate? I think that's fine given the additional context you added to ``to_dict``, I just wanted to check that I'm not missing anything.

Reply:
I think you're reviewing an old commit, try again and make sure to "clear filters" at the top?

Reply:
Hmm I'm not sure what I missed. I always review all the commits since the last review, and check the full list of files. Sometimes GitHub caches things and doesn't update them without a page refresh, though. :/

This does seem resolved.